storage/spanstore/downsampling_writer.go

```diff
@@ -0,0 +1,135 @@
+	SpansDropped  metrics.Counter `metric:"spans_dropped"`
+	SpansAccepted metrics.Counter `metric:"spans_accepted"`
+}
+
+// DownSamplingWriter is a span Writer that drops spans with a predefined downSamplingRatio.
+type DownSamplingWriter struct {
+	spanWriter   Writer
+	threshold    uint64
+	lengthOfSalt int
+	poolInstance poolSingleton
+	metrics      *downSamplingWriterMetrics
+}
+
+// DownSamplingOptions contains the options for constructing a DownSamplingWriter.
+type DownSamplingOptions struct {
+	Ratio          float64
+	HashSalt       string
+	MetricsFactory metrics.Factory
+}
+
+// Initialize the singleton if it doesn't exist already.
+func getHashPoolInstance(length int, hashSaltBytes []byte) poolSingleton {
+	once.Do(func() {
+		instance = poolSingleton{
+			hashPool: &sync.Pool{
+				New: func() interface{} {
+					return fnv.New64a()
+				},
+			},
+			bytePool: &sync.Pool{
+				New: func() interface{} {
+					// Since Get selects an arbitrary item from the Pool, removes it from the
+					// Pool, and returns it to the caller and it may choose to ignore the pool
+					// and treat it as empty, hashSaltBytes needs to be written to byteArray at initialization
+					byteArray := make([]byte, length)
+					copy(byteArray[:len(hashSaltBytes)], hashSaltBytes)
+					return &byteArray
+				},
+			},
+		}
+	})
+	return instance
+}
+
+// NewDownSamplingWriter creates a DownSamplingWriter.
+func NewDownSamplingWriter(spanWriter Writer, downSamplingOptions DownSamplingOptions) *DownSamplingWriter {
+	threshold := uint64(downSamplingOptions.Ratio * float64(math.MaxUint64))
+	writeMetrics := &downSamplingWriterMetrics{}
+	metrics.Init(writeMetrics, downSamplingOptions.MetricsFactory, nil)
+	hashSaltBytes := []byte(downSamplingOptions.HashSalt)
+	poolInstance := getHashPoolInstance(len(hashSaltBytes)+traceIDByteSize, hashSaltBytes)
+	return &DownSamplingWriter{
+		spanWriter:   spanWriter,
+		threshold:    threshold,
+		poolInstance: poolInstance,
+		metrics:      writeMetrics,
+		lengthOfSalt: len(hashSaltBytes),
+	}
+}
+
+// WriteSpan calls WriteSpan on wrapped span writer.
+func (ds *DownSamplingWriter) WriteSpan(span *model.Span) error {
+	if ds.shouldDownsample(span) {
+		// Drops spans when hashVal falls beyond computed threshold.
+		ds.metrics.SpansDropped.Inc(1)
+		return nil
+	}
+	ds.metrics.SpansAccepted.Inc(1)
+	return ds.spanWriter.WriteSpan(span)
+}
+
+func (ds *DownSamplingWriter) shouldDownsample(span *model.Span) bool {
+	byteSlice := ds.poolInstance.bytePool.Get().(*[]byte)
+	// Currently MarshalTo will only return err if size of traceIDBytes is smaller than 16
+	// Since we force traceIDBytes to be size of 16 metrics is not necessary here.
+	span.TraceID.MarshalTo((*byteSlice)[ds.lengthOfSalt:])
+	hashVal := ds.hashBytes(*byteSlice)
+	ds.poolInstance.bytePool.Put(byteSlice)
+	return hashVal >= ds.threshold
+}
+
+// hashBytes returns the uint64 hash value of byte slice.
+func (ds *DownSamplingWriter) hashBytes(bytes []byte) uint64 {
```

Comment:
once you combine hash and buffer into `hasher` struct, you can define this method on that struct instead of on the Writer